torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+// index hoisting. For example, if I have a loop structure and index:
+//   FOR i1
+//     FOR i2
+//       FOR i3
+//         index = ((i3 + i2) + i1) + 256
+// There is no hoisting opportunity for this index in this loop structure.
+// However, if I transform the index into index = ((256 + i1) + i2) + i3,
+// then I can hoist the index as
+//   FOR i1
+//     i4 = (256 + i1)
+//     FOR i2
+//       i5 = i4 + i2
+//       FOR i3
+//         index = i5 + i3
+// This minimizes the total number of computations.
+
+bool isAssociativeAndCommutative(BinaryOpType type) {
+  return type == BinaryOpType::Add || type == BinaryOpType::Mul ||
+      type == BinaryOpType::And || type == BinaryOpType::Or ||
+      type == BinaryOpType::Xor;
+}
+
+// The expression type that represents the flattened ops. For example, if I have
+// out = a + b + 3 + c + 5, then I will have:
+//   FlattenedAssocCommOp {
+//     inputs: [a, b, c]
+//     putputs: [out]
+//     constant term: 8
+//   }
+//
+// TODO: refactor IR printing so that private exprs like this can be printed.
+// This will be very helpful for debugging.
+class FlattenedAssocCommOp : public Expr {
+ public:
+  using Expr::Expr;
+
+  FlattenedAssocCommOp(
+      IrBuilderPasskey passkey,
+      BinaryOpType op,
+      Val* out,
+      std::vector<Val*> terms)
+      : Expr(passkey) {
+    TORCH_CHECK(
+        isAssociativeAndCommutative(op),
+        "Can only flatten associative and commutative ops");
+    addAttribute(
+        IrBuilder::create<Attribute<BinaryOpType>>(passkey.ir_container_, op));
+    addOutput(out);
+    for (auto v : terms) {
+      // Note that `addInput` is overriden in this class.
+      addInput(v);
+    }
+  }
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual const char* getOpString() const override {
+    return "FlattenedAssocCommOp";
+  }
+
+  DataType dtype() const {
+    return *output(0)->getDataType();
+  }
+
+  BinaryOpType getOpType() const {
+    return attribute(0)->as<Attribute<BinaryOpType>>()->value;
+  }
+
+  // Constant terms are stored and handled separately from other inputs. For
+  // example, if I have a + b + 3 + c + 5, then the constant term will be 8, and
+  // a, b, c are all inputs.
+  bool hasConstantTerm() const {
+    return attributes().size() == 2;
+  }
+
+  Val* getConstantTerm() const {
+    return attributeVal(1);
+  }
+
+  template <typename T>
+  T getConstantTermValue() const {
+    return *getConstantTerm()->as<Scalar<T>>()->value();
+  }
+
+  // Get a vector of inputs, sorted as the order given by `variables`. Note that
+  // the sorting key is the rightmost variable that an input depends on. For
+  // example, if I have two inputs.
+  // v1 = a * c
+  // v2 = b
+  // and variables is [a, b, c], then v2 < v1 because the rightmost depending
+  // variable of v2 is b, and the rightmost depending variable of v1 is c,
+  // and b < c. So in this example, this function will return [v2, v1].
```

Comment:
It took me a little while to get this, the variable list is setting the order of the dependencies of v1 and v2. So you have inputs on the expression, and you have an ordering of variables specified. You find all variables in the variable list that each input depends on.

So when you loop over variables, if a variable is found in one input but not in the other, you'll put that input to the right of the other. So the input that exclusively has the right most variable in variable list will be to the right of the other input.

Reply:
Yes, your understanding is correct. I copy-pasted part of your reply here to the code as a comment.